theano/contrib/mkl/mkl_ndarray.c

```diff
@@ -0,0 +1,1361 @@
+ *
+ */
+static PyObject*
+MKLNdarray_get_ndim(MKLNdarray *self, void *closure) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_get_ndim: input MKLNdarray* self is NULL");
+        return NULL;
+    } else {
+        return PyInt_FromLong(self->nd);
+    }
+}
+
+
+/*
+ * Get size info of a MKLNdarray instance.
+ *
+ * Register in MKLNdarray_getset
+ *
+ * Return a integer number.
+ *
+ */
+static PyObject*
+MKLNdarray_get_size(MKLNdarray *self, void *closure) {
+    size_t total_element = 1;
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_get_size: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    if (self->nd <= 0) {
+        total_element = 0;
+    } else {
+        for (int i = 0; i < self->nd; i++) {
+            total_element *= self->user_structure[i];
+        }
+    }
+    return PyInt_FromLong(total_element);
+}
+
+
+/*
+ * Get base info of a MKLNdarray instance.
+ *
+ * Register in MKLNdarray_getset
+ *
+ * Return a PyObject.
+ *
+ */
+static PyObject*
+MKLNdarray_get_base(MKLNdarray *self, void *closure) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_get_base: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    PyObject * base = self->base;
+    if (!base) {
+        base = Py_None;
+    }
+    Py_INCREF(base);
+    return base;
+}
+
+
+/*
+ * Create a PyArrayObject from a MKLNdarray.
+ */
+PyObject* MKLNdarray_CreateArrayObj(MKLNdarray *self) {
+    if (self == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_CreateArrayObj: input MKLNdarray* self is NULL");
+        return NULL;
+    }
+
+    if (self->nd < 0 ||
+        self->private_data == NULL ||
+        self->private_layout == NULL) {
+        PyErr_SetString(PyExc_RuntimeError,
+                        "MKLNdarray_CreateArrayObj: Can't convert from an uninitialized MKLNdarray");
+        return NULL;
+    }
+
+    npy_intp npydims[MAX_NDIM] = {0};
+    for (int i = 0; i < self->nd; i++) {
+        npydims[i] = (npy_intp)self->user_structure[i];
+    }
+
+    PyArrayObject* rval = NULL;
+    if (self->dtype == MKL_FLOAT64) {
+        // float64
+        rval = (PyArrayObject*)PyArray_SimpleNew(self->nd, npydims, NPY_FLOAT64);
+    } else {
+        // float32
+        rval = (PyArrayObject*)PyArray_SimpleNew(self->nd, npydims, NPY_FLOAT32);
+    }
+
+    if (!rval) {
+        return NULL;
+    }
+
+    void* rval_data = PyArray_DATA(rval);
+    dnnLayout_t layout_user = NULL;
+    int status = -1;
+    dnnPrimitive_t primitive = NULL;
+
+    size_t mkl_size[MAX_NDIM] = {0};
+    size_t mkl_stride[MAX_NDIM] = {0};
+
+    // nchw -> whcn
+    for (int i = 0; i < self->nd; i++) {
+        mkl_size[i] = (MKLNdarray_DIMS(self))[self->nd - i - 1];
+        mkl_stride[i] = (MKLNdarray_STRIDES(self))[self->nd - i -1];
```

Comment:
MKL object has its own description for the layout, and it's enough to MKL object.
The reason to keep the original information (strides and dims) is that, when we convert MKL object to Numpy Ndarray we need to know the layout information of target object (Numpy Ndarray).